lib/moab/storage_object_validator.rb

```diff
@@ -88,18 +100,70 @@ def check_correctly_formed_moabs
       version_directories.each do |version_dir|
         version_path = "#{storage_obj_path}/#{version_dir}"
         version_sub_dirs = sub_dirs(version_path)
-        before_result_size = errors.size
-        errors.concat check_sub_dirs(version_sub_dirs, version_dir, EXPECTED_VERSION_SUB_DIRS)
-        after_result_size = errors.size
+        version_error_count = errors.size
+        errors.concat check_version_sub_dirs(version_sub_dirs, version_dir)
         # run the following checks if this version dir passes check_sub_dirs, even if some prior version dirs didn't
-        if before_result_size == after_result_size
-          data_dir_path = "#{version_path}/#{DATA_DIR_NAME}"
-          data_sub_dirs = sub_dirs(data_dir_path)
-          errors.concat check_sub_dirs(data_sub_dirs, version_dir, EXPECTED_DATA_SUB_DIRS)
-          errors.concat check_required_manifest_files(version_path, version_dir)
+        errors.concat check_required_manifest_files(version_path, version_dir) if version_error_count == errors.size
+        errors.concat check_data_sub_dirs(version_path, version_dir) if version_error_count == errors.size
+        errors.concat check_metadata_dir_files_only(version_path) if version_error_count == errors.size
+      end
+      errors
+    end
+
+    def check_version_sub_dirs(version_sub_dirs, version)
+      errors = []
+      version_sub_dir_count = version_sub_dirs.size
+      if version_sub_dir_count == EXPECTED_VERSION_SUB_DIRS.size
+        errors.concat expected_dirs(version_sub_dirs, version, EXPECTED_VERSION_SUB_DIRS)
+      elsif version_sub_dir_count > EXPECTED_VERSION_SUB_DIRS.size
+        errors.concat found_unexpected(version_sub_dirs, version, EXPECTED_VERSION_SUB_DIRS)
+      elsif version_sub_dir_count < EXPECTED_VERSION_SUB_DIRS.size
+        errors.concat missing_dir(version_sub_dirs, version, EXPECTED_VERSION_SUB_DIRS)
+      end
+      errors
+    end
+
+    def check_data_sub_dirs(version_path, version)
+      errors = []
+      data_dir_path = "#{version_path}/#{DATA_DIR}"
+      data_sub_dirs = sub_dirs(data_dir_path)
+      if data_sub_dirs.size > EXPECTED_DATA_SUB_DIRS.size
+        errors.concat found_unexpected(data_sub_dirs, version, EXPECTED_DATA_SUB_DIRS)
+      else
+        errors.concat missing_dir(data_sub_dirs, version, [METADATA_DIR]) unless data_sub_dirs.include?(METADATA_DIR)
+        errors.concat found_unexpected(data_sub_dirs, version, EXPECTED_DATA_SUB_DIRS) unless subset?(data_sub_dirs,
+                                                                                                      EXPECTED_DATA_SUB_DIRS)
+
+        if data_sub_dirs.include?('content')
```

Comment:
you should either put the metadata_dir_files check in here, or pull the content dir files check out of here -- consistency is appreciated.

Reply:
👍